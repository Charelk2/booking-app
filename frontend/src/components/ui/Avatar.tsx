"use client";
import Image from "next/image";

interface AvatarProps {
  profileUrl?: string;
  size?: number;
  alt?: string;
}

export default function Avatar({
  profileUrl,
  size = 48,
  alt = "avatar",
}: AvatarProps) {
  const src = profileUrl
    ? `${process.env.NEXT_PUBLIC_API_URL}/static/profile_pics/${profileUrl}`
    : "/static/default-avatar.svg";

  return (
    <Image
      src={src}
      width={size}
      height={size}
      className="object-cover rounded-full"
      alt={alt}
    />
  );
}
