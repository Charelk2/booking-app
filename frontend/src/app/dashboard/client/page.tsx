"use client";

import React, { useEffect, useState, useMemo } from "react";
import { useRouter, usePathname, useSearchParams } from "next/navigation";
import MainLayout from "@/components/layout/MainLayout";
import { useAuth } from "@/contexts/AuthContext";
import { Booking, BookingRequest } from "@/types";
import { getMyClientBookings, getMyBookingRequests } from "@/lib/api";
import SectionList from "@/components/dashboard/SectionList";
import BookingRequestCard from "@/components/dashboard/BookingRequestCard";
import { Spinner } from "@/components/ui";
import DashboardTabs from "@/components/dashboard/DashboardTabs";
import { format } from "date-fns";
import { formatCurrency, formatStatus } from "@/lib/utils";
import Link from "next/link";

export default function ClientDashboardPage() {
  const { user, loading: authLoading } = useAuth();
  const router = useRouter();
  const pathname = usePathname();
  const params = useSearchParams();

  const [bookings, setBookings] = useState<Booking[]>([]);
  const [bookingRequests, setBookingRequests] = useState<BookingRequest[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState("");

  const initialTab = params.get("tab") === "bookings" ? "bookings" : "requests";
  const [activeTab, setActiveTab] = useState<"requests" | "bookings">(initialTab);

  useEffect(() => {
    const params = new URLSearchParams(window.location.search);
    params.set("tab", activeTab);
    router.replace(`${pathname}?${params.toString()}`);
  }, [activeTab, router, pathname]);

  useEffect(() => {
    if (authLoading) return;
    if (!user) {
      router.push(`/login?next=${encodeURIComponent(pathname)}`);
      return;
    }
    if (user.user_type !== "client") {
      router.push("/dashboard/artist");
      return;
    }

    const fetchData = async () => {
      try {
        const [bookingsData, requestsData] = await Promise.all([
          getMyClientBookings(),
          getMyBookingRequests(),
        ]);
        setBookings(bookingsData.data);
        setBookingRequests(requestsData.data);
      } catch (err) {
        console.error("Client dashboard fetch error:", err);
        setError("Failed to load dashboard data. Please try again.");
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [user, authLoading, router, pathname]);

  const upcomingBookings = useMemo(() => {
    const now = new Date().getTime();
    return bookings
      .filter((b) => new Date(b.start_time).getTime() >= now)
      .sort((a, b) => new Date(a.start_time).getTime() - new Date(b.start_time).getTime())
      .slice(0, 5);
  }, [bookings]);

  if (loading) {
    return (
      <MainLayout>
        <div className="p-8 flex justify-center"><Spinner /></div>
      </MainLayout>
    );
  }

  if (error) {
    return (
      <MainLayout>
        <div className="p-8 text-red-600">{error}</div>
      </MainLayout>
    );
  }

  return (
    <MainLayout>
      <div className="max-w-5xl mx-auto px-4 py-8">
        <DashboardTabs activeTab={activeTab} onChange={setActiveTab} showServices={false} />
        <div className="mt-6">
          {activeTab === "requests" && (
            <section>
              <SectionList
                title="Booking Requests"
                data={bookingRequests}
                renderItem={(r) => <BookingRequestCard key={r.id} request={r} onUpdate={() => {}} />}
                emptyState={<span>No requests yet</span>}
              />
            </section>
          )}
          {activeTab === "bookings" && (
            <section>
              <SectionList
                title="Upcoming Bookings"
                data={upcomingBookings}
                emptyState={<span>No bookings yet</span>}
                renderItem={(booking) => (
                  <div key={booking.id} className="bg-white p-4 shadow rounded-lg">
                    <div className="font-medium text-gray-900">
                      {booking.artist?.first_name || "Unknown"} {booking.artist?.last_name || ""}
                    </div>
                    <div className="text-sm text-gray-500">{booking.service?.title || "â€”"}</div>
                    <div className="text-sm text-gray-500">
                      {format(new Date(booking.start_time), "MMM d, yyyy h:mm a")}
                    </div>
                    <div className="mt-2 flex justify-between items-center">
                      <span
                        className={`inline-flex rounded-full px-2 text-xs font-semibold leading-5 ${
                          booking.status === "completed"
                            ? "bg-green-100 text-green-800"
                            : booking.status === "cancelled"
                            ? "bg-red-100 text-red-800"
                            : booking.status === "confirmed"
                            ? "bg-brand-light text-brand-dark"
                            : "bg-yellow-100 text-yellow-800"
                        }`}
                      >
                        {formatStatus(booking.status)}
                      </span>
                      <span className="text-sm text-gray-500">
                        {formatCurrency(Number(booking.total_price))}
                      </span>
                    </div>
                  </div>
                )}
              />
              {bookings.length > upcomingBookings.length && (
                <div className="mt-2">
                  <Link href="/dashboard/client/bookings" className="text-brand-dark hover:underline text-sm">
                    View All Bookings
                  </Link>
                </div>
              )}
            </section>
          )}
        </div>
      </div>
    </MainLayout>
  );
}

