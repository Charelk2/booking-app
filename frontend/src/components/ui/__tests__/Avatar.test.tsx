import { createRoot } from 'react-dom/client';
import { act } from 'react-dom/test-utils';
import React from 'react';
import Avatar from '../Avatar';

describe('Avatar component', () => {
  let container: HTMLDivElement;
  let root: ReturnType<typeof createRoot>;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    root = createRoot(container);
  });

  afterEach(() => {
    act(() => {
      root.unmount();
    });
    container.remove();
  });

  it('matches snapshot for default placeholder', () => {
    act(() => {
      root.render(<Avatar />);
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with initials', () => {
    act(() => {
      root.render(<Avatar initials="A" />);
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
