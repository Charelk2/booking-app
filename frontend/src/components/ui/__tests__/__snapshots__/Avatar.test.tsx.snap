// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar component matches snapshot for default placeholder 1`] = `
<div
  class="flex-shrink-0 rounded-full bg-indigo-100 flex items-center justify-center text-gray-700 font-medium overflow-hidden"
  style="width: 40px; height: 40px;"
>
  <img
    alt="avatar"
    class="object-cover rounded-full"
    data-nimg="1"
    decoding="async"
    height="40"
    loading="lazy"
    src="http://localhost:8000/static/default-avatar.svg"
    style="color: transparent;"
    width="40"
  />
</div>
`;

exports[`Avatar component matches snapshot with initials 1`] = `
<div
  class="flex-shrink-0 rounded-full bg-indigo-100 flex items-center justify-center text-gray-700 font-medium overflow-hidden"
  style="width: 40px; height: 40px;"
>
  <span>
    A
  </span>
</div>
`;
